go/jsonclient/client.go

```diff
@@ -0,0 +1,191 @@
+		hc = new(http.Client)
+	}
+	return &JSONClient{uri: strings.TrimRight(uri, "/"), httpClient: hc, Verifier: verifier}, nil
+}
+
+// GetAndParse makes a HTTP GET call to the given path, and attempt to parse
+// the response as a JSON representation of the rsp structure.  The provided
+// context is used to control the HTTP call.
+func (c *JSONClient) GetAndParse(ctx context.Context, path string, params map[string]string, rsp interface{}) (*http.Response, error) {
+	if ctx == nil {
+		return nil, errors.New("Context required")
+	}
+	// Build a GET request with URL-encoded parameters.
+	vals := url.Values{}
+	for k, v := range params {
+		vals.Add(k, v)
+	}
+	fullURI := fmt.Sprintf("%s%s?%s", c.uri, path, vals.Encode())
+	httpReq, err := http.NewRequest(http.MethodGet, fullURI, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	httpRsp, err := ctxhttp.Do(ctx, c.httpClient, httpReq)
+	if err != nil {
+		return nil, err
+	}
+	// Make sure everything is read, so http.Client can reuse the connection.
+	defer httpRsp.Body.Close()
+	defer ioutil.ReadAll(httpRsp.Body)
+
+	if httpRsp.StatusCode != http.StatusOK {
+		return httpRsp, fmt.Errorf("got HTTP Status %s", httpRsp.Status)
+	}
+
+	if err := json.NewDecoder(httpRsp.Body).Decode(rsp); err != nil {
+		return httpRsp, err
+	}
+
+	return httpRsp, nil
+}
+
+// PostAndParse makes a HTTP POST call to the given path, including the request
+// parameters, and attempt to parse the response as a JSON representation of
+// the rsp structure.  The provided context is used the control the HTTP call.
+func (c *JSONClient) PostAndParse(ctx context.Context, path string, req interface{}, rsp interface{}) (*http.Response, error) {
+	if ctx == nil {
+		return nil, errors.New("Context required")
+	}
+	// Build a POST request with JSON body.
+	postBody, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+	fullURI := fmt.Sprintf("%s%s", c.uri, path)
+	httpReq, err := http.NewRequest(http.MethodPost, fullURI, bytes.NewReader(postBody))
+	if err != nil {
+		return nil, err
+	}
+	httpReq.Header.Set("Content-Type", "application/json")
+
+	httpRsp, err := ctxhttp.Do(ctx, c.httpClient, httpReq)
+
+	// Read all of the body, if there is one, so that the http.Client can do Keep-Alive.
+	var body []byte
+	if httpRsp != nil {
+		body, err = ioutil.ReadAll(httpRsp.Body)
+		httpRsp.Body.Close()
+	}
+	if err != nil {
+		return httpRsp, err
+	}
+	if httpRsp.StatusCode == http.StatusOK {
+		if err = json.Unmarshal(body, &rsp); err != nil {
+			return httpRsp, err
+		}
+	}
+	return httpRsp, nil
+}
+
+// PostAndParseWithRetry makes a HTTP POST call, but retries (with backoff) on
+// retriable errors.
+func (c *JSONClient) PostAndParseWithRetry(ctx context.Context, path string, req interface{}, rsp interface{}) (*http.Response, error) {
+	if ctx == nil {
+		return nil, errors.New("Context required")
+	}
+	httpStatus := "Unknown"
+	// Retry after 1s, 2s, 4s, 8s, 16s, 32s, 64s, 128s, 128s, ....
+	maxInterval := 128.0
```

Comment:
curious why these are all floats - I'd have thought integers would be fine?  also, how about using time.Duration?